### PR TITLE
allow metrictank to run as graphite 1.0 cluster node

### PR DESCRIPTION
**Pull request by woodsaj**
Allow metrictank to sit behind a stock graphite-web 1.0.0 deployment
as a cluster_server.  This allows metrictank to be used without the
need of using a custom finder plugin.

To support this feature the following changes have been made
- support pickle format  for metrics/find calls
- support pickle format for render calls

### CODE REVIEW

**Comment by DanCech** [comment]
https://pypi.python.org/pypi/graphite-web/1.0.1

File: docker/docker-dev-direct-single-tenant/datasources/mt-graphite-api [hunk 1]
```diff
@@ -1,7 +0,0 @@
-{
-  "name":"mt-graphite-api",
-  "type":"graphite",
-  "url":"http://localhost:8080",
-  "access":"direct",
-  "isDefault":true
-}
```
Comment (Dieterbe) [inline docker/docker-dev-direct-single-tenant/datasources/mt-graphite-api]: so we remove the mt-graphite-api DS but why don't we add a mt-graphite DS?

File: docker/docker-dev/docker-compose.yml [hunk 2]
```diff
@@ -35,6 +35,9 @@ services:
      - "8080:80"
     environment:
       GRAPHITE_CLUSTER_SERVERS: metrictank:6060
+      WSGI_PROCESSES: 4
+      WSGI_THREADS: 25
+      GRAPHITE_STATSD_HOST: statsdaemon
```
Comment (woodsaj) [inline docker/docker-dev/docker-compose.yml]: GRAPHITE_* are graphite settings, defined in graphite's local_settings.py
WSGI_* are wsgi settings, defined in the WSGIDaemonProcess directive in Apache.

File: docs/graphite.md [hunk 3]
```diff
@@ -1,4 +1,4 @@
-# Graphite-api
+# Graphite
```
Comment (Dieterbe) [inline docs/graphite.md]: please run a `grep -Ri graphite-api` in the root of the MT repo.  you'll find plenty of places that need to be updated still. like the documentation index in the readme,

File: api/models/graphite.go [hunk 4]
```diff
@@ -15,7 +15,8 @@ type GraphiteRender struct {
 	From          string   `json:"from" form:"from"`
 	Until         string   `json:"until" form:"until"`
 	To            string   `json:"to" form:"to"`
-	Format        string   `json:"format" form:"format" binding:"In(,json,msgp)"`
+	Format        string   `json:"format" form:"format" binding:"In(,json,msgp,pickle)"`
+	LocalOnly     bool     `json:"local" form:"local"` //this is set to true by graphite-web when it passes request to cluster servers
```
Comment (woodsaj) [inline api/models/graphite.go]: graphite should only send queries with all functions stripped from the target param. (ie, it just sends `target=<pathExpr>&local=true`)

> "local" is ambiguous here.

Nothing we can do about that.  "local" is the query param that graphite sends.

The only reason we handle this query param is to prevent an infinite loop if either graphite or metrictank incorrectly parse the target.  if graphite sends a target with a pathExpr wrapped in a function, or MT thinks the target cant be processed locally and needs to be proxied, bad things will happen.

We can change the stuct field to "MetrictankOnly" if that helps.

File: docs/installation-deb.md [hunk 5]
```diff
@@ -25,7 +23,7 @@ instance or a cluster.  Metrictank will also respond to queries: if the data is
 RAM, and older data is fetched from cassandra.  This happens transparantly.
 Metrictank maintains an index of metrics metadata, for all series it sees.  
 You can use an index entirely in memory, or backed by Cassandra for persistence.
-You'll typically query metrictank by querying graphite-api which uses the graphite-metrictank plugin to talk
+You'll typically query metrictank by querying graphite which uses the clustering capabilities of graphite to talk
 to metrictank.  You can also query metrictank directly but this is experimental and too early for anything useful.
```
Comment (Dieterbe) [inline docs/installation-deb.md]: "You can also query metrictank directly but this is experimental and too early for anything useful." this is not true anymore. in fact querying MT directly is now the preferred way. Let's replace this sentence, along with the "You'll typically..." sentence by something like: "You can query metrictank directly (it has fast, but limited built-in processing and will fallback to graphite when needed) or you can also just query graphite which will always use graphite's processing but use metrictank as a datastore" except I'm not sure if the latter statement is true.

File: docs/quick-start-docker.md [hunk 6]
```diff
@@ -31,7 +31,7 @@ The stack will listen on the following ports:
 * 2003 tcp (metrictank's carbon input)
 * 3000 tcp (grafana's http port)
 * 6060 tcp (metrictank's internal endpoint)
-* 8080 tcp (the graphite api query endpoint)
+* 8080 tcp (the graphite query endpoint)
```
Comment (Dieterbe) [inline docs/quick-start-docker.md]: you changed it to port 80

File: docs/quick-start-docker.md [hunk 7]
```diff
@@ -85,14 +85,13 @@ Add a new data source with with these settings:
 * Name: `metrictank`
 * Default: `yes`
 * Type: `Graphite`
-* Url: `http://localhost:8080`
-* Access: `direct` (not `proxy`)
+* Url: `http://graphite`
+* Access: `proxy`
```
Comment (woodsaj) [inline docs/quick-start-docker.md]: graphite-web does not support CORS. so direct does not work.

users can still connect directly to the datasource on localhost:8080.

File: docs/installation-deb.md [hunk 8]
```diff
@@ -4,11 +4,9 @@
 
 We'll go over these in more detail below.
 
-* Cassandra. We run and recommend 3.0.8 .
+* Cassandra. We run and recommend 3.X .
```
Comment (Dieterbe) [inline docs/installation-deb.md]: can't we be more specific? 3.x includes old versions that don't have some things we need (like TWCS)

File: docs/quick-start-docker.md (hunk 6 above)
Comment (woodsaj) [inline docs/quick-start-docker.md]: the container listens on 80, but it is bound to port 8080 on localhost.

File: docs/installation-deb.md [hunk 9]
```diff
@@ -81,9 +53,39 @@ Supported distributions:
...
 
 
+### Install Graphite
+
+Install Graphite via your prefered method as detailed at http://graphite.readthedocs.io/en/latest/install.html
+(We hope to provide Debian and Ubuntu packages in the near future.)
+
+Configure graphite with the following settings in local_settings.py
+```
+CLUSTER_SERVERS = ['localhost:6060'] # update to match the host:port metrictank is running on.
```
Comment (Dieterbe) [inline docs/installation-deb.md]: at this point in the guide, MT is not running yet, and the user has no idea yet about how and on which port he will run MT, as that's only at the end of the guide

File: docs/installation-deb.md [hunk 10]
```diff
@@ -115,7 +117,7 @@ The log - should you need it - is at /var/log/cassandra/system.log
 
 ## Set up statsd
 
-You can optionally statsd or a statsd-compatible agent for instrumentation of graphite-api.
+You can optionally statsd or a statsd-compatible agent for instrumentation of your applications.
```
Comment (Dieterbe) [inline docs/installation-deb.md]: too vague. maybe make it "of graphite and optionally any of your other applications"

File: docs/installation-deb.md (hunk 8 above)
Comment (woodsaj) [inline docs/installation-deb.md]: yep you are correct. I wrongly assumed all 3.X versions had twcs, but it was not added till 3.8

File: docs/installation-deb.md (hunk 9 above)
Comment (woodsaj) [inline docs/installation-deb.md]: this is still a big improvement on the previous instructions, which didnt even mention configuring graphite-api.

refactoring this install guide is outside the scope of this PR.

File: docs/quick-start-docker.md [hunk 11]
```diff
@@ -85,14 +85,13 @@ Add a new data source with with these settings:
...
 
 When you hit save, Grafana should succeed in talking to the data source.
 
 ![Add data source screenshot](https://raw.githubusercontent.com/raintank/metrictank/master/docs/assets/add-datasource-docker.png)
 
-Note: it also works with `proxy` mode but then you have to enter `http://graphite-api:8080` as uri.
```
Comment (Dieterbe) [inline docs/quick-start-docker.md]: this would be a good place to mention that users can still connect directly to the datasource on localhost:8080.

File: docs/installation-deb.md (hunk 9 above)
Comment (Dieterbe) [inline docs/installation-deb.md]: all i'm saying is you're introducing a sentence that is confusing. in fact, the rest of the guide assumes mt will run on localhost:6060 so let's remove the comment, as it instructs people to do something that they don't know about, and they don't need to do it anyway.

File: docker/extra/dashboards/graphite.json [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+              "refId": "A",
+              "target": "alias(averageSeries(stats.docker-env.timers.view.graphite.render.views.renderView.*.mean), 'mean')"
+            },
+            {
+              "refId": "B",
+              "target": "alias(maxSeries(stats.docker-env.timers.view.graphite.render.views.renderView.*.upper_90), 'upper_90')"
+            },
+            {
+              "refId": "C",
+              "target": "alias(maxSeries(stats.docker-env.timers.view.graphite.render.views.renderView.*.upper), 'upper')"
```
Comment (woodsaj) [inline docker/extra/dashboards/graphite.json]: this dashboard cant be use anywhere else.

File: api/models/graphite.go (hunk 4 above)
Comment (Dieterbe) [inline api/models/graphite.go]: yeah MTOnly or NoProxy or something

File: docker/docker-dev/docker-compose.yml (hunk 2 above)
Comment (Dieterbe) [inline docker/docker-dev/docker-compose.yml]: what i mean is, the reason behind the different docker environments, is that they all have something unique to them, and usually relating to MT. eg: 
* docker-standard : standard docker env. 1 graphite, 1 MT, official binaries and stock configs
* docker-dev: same, but custom MT binary and MT config files mounted in (though the defaults are what is in git, so equivalent to docker-standard by default, which makes it easy to use `git diff` to see the difference)
* docker-cluster : a cluster of MT's

So I don't think we should commit any changes that make the graphite in docker-dev behave any different from any other of the docker environments. maybe those settings be set so that they apply to all our docker environments

File: docker/docker-dev-direct-single-tenant/datasources/mt-graphite-api (hunk 1 above)
Comment (woodsaj) [inline docker/docker-dev-direct-single-tenant/datasources/mt-graphite-api]: i removed docker/docker-dev-direct-single-tenant completely, as it is no different to docker-dev.
All docker envs that use direct carbon input are configured as single-tenant (as all metrics ingested use orgId=1).  the cluster envs, use multiTenant metrictank, but have graphite configured in singleTenant mode so it sends the x-org-id:1 header, preventing the need for running tsdb-gw.

Comment (Dieterbe) [inline docker/docker-dev-direct-single-tenant/datasources/mt-graphite-api]: ok that sounds good.  but I still think it's useful to have 2 datasources (one to talk to graphite, and one to talk to MT). maybe not in the standard one, but definitely in docker-dev.

Comment (woodsaj) [inline docker/docker-dev-direct-single-tenant/datasources/mt-graphite-api]: agreed. I ensured all envs have a graphite and metrictank datasource, with the metrictank one set as default.

File: docs/quick-start-docker.md [hunk 13]
```diff
@@ -102,7 +101,6 @@ In the metrics tab you should see a bunch of metrics already in the root hierarc
 * `metrictank.usage`: usage metrics reported by metrictank.  See
 [Usage reporting](https://github.com/raintank/metrictank/blob/master/docs/usage-reporting.md)
 It may take a few minutes for the usage metrics to show up.
-* `stats`: these are metrics coming from graphite-api, aggregated by statsdaemon and sent back to metrictank every second.
```
Comment (Dieterbe) [inline docs/quick-start-docker.md]: we should add a note here that this won't show up until statsdaemon receives some stats, there's only a few non-obvious cases in which this happens (user adds graphite-direct datasource, user sends to statsdaemon manually, or user does requests to MT which MT needs graphite for)

File: api/response/pickle_test.go [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+			Datapoints: points,
+			Interval:   10,
+		},
+	}
+	b.SetBytes(int64(len(points) * 12))
+
+	b.ResetTimer()
+	var resp *Pickle
+	for n := 0; n < b.N; n++ {
+		resp = NewPickle(200, models.SeriesPickleFormat(data))
```
Comment (Dieterbe) [inline api/response/pickle_test.go]: i added the benchmark.

**Comment by Dieterbe** [comment]
sweet, getting much better perf out of graphite.
just one minor thing left, https://github.com/raintank/metrictank/pull/611#discussion_r113076381, otherwise, this LGTM

File: docs/quick-start-docker.md [hunk 15]
```diff
@@ -102,8 +103,7 @@ In the metrics tab you should see a bunch of metrics already in the root hierarc
...
-
+* `stats`: these are metrics coming from gaphite, aggregated by statsdaemon and sent back to metrictank every second.
```
Comment (DanCech) [inline docs/quick-start-docker.md]: s/gaphite/graphite/

**Comment by woodsaj** [comment]
thanks @Dieterbe all issues raised have been addressed.  please approve an LGTM
